### ISSUE TITLE
Invalid control flow after EmitMeshTasksEXT()

### ISSUE THREAD

**Issue by HansKristian-Work** [KhronosGroup/glslang#3020]
This shader hits a bug:

```
#version 450
#extension GL_EXT_mesh_shader : require
layout(local_size_x = 4, local_size_y = 3, local_size_z = 2) in;

struct Payload
{
	float v[3];
};
taskPayloadSharedEXT Payload p;
shared float vs[24];

void main()
{
	vs[gl_LocalInvocationIndex] = 10.0;
	barrier();
	if (gl_LocalInvocationIndex < 12)
		vs[gl_LocalInvocationIndex] += vs[gl_LocalInvocationIndex + 12];
	barrier();
	if (gl_LocalInvocationIndex < 6)
		vs[gl_LocalInvocationIndex] += vs[gl_LocalInvocationIndex + 6];
	barrier();
	if (gl_LocalInvocationIndex < 3)
		vs[gl_LocalInvocationIndex] += vs[gl_LocalInvocationIndex + 3];
	barrier();

	p.v[gl_LocalInvocationIndex] = vs[gl_LocalInvocationIndex];
	if (vs[5] > 20.0)
	{
		EmitMeshTasksEXT(int(vs[4]), int(vs[6]), int(vs[8]));
	}
	else
	{
		EmitMeshTasksEXT(int(vs[6]), 10, 50u);
	}
}
```

```
              94:    6(float)   Load 93
              95:     67(int)   ConvertFToS 94
              96:      7(int)   Bitcast 95
                                EmitMeshTasksEXT 86 91 96 66(p)
                                Branch 81                                <----- *OpEmitMeshTasksEXT* must be the last instruction in a block.
              97:               Label
              98:     16(ptr)   AccessChain 11(vs) 87
              99:    6(float)   Load 98
             100:     67(int)   ConvertFToS 99
             101:      7(int)   Bitcast 100
                                EmitMeshTasksEXT 101 102 103 66(p)
                                Branch 81
```

According to SPIR-V spec, this is invalid. https://github.com/KhronosGroup/SPIRV-Registry/pull/165/files#diff-9d8cf769182417e1989cb51f487a381c2f25f4a714b010b25318dc59578e23ccR160

**Comment by pmistryNV** [comment]
From the SPIR-V spec "Behavior is undefined if any invocation terminates without executing this instruction, or if any invocation executes this instruction in non-uniform control flow. " This is beacuse it's hard to detect non uniform control flow in compiler statically.

**Comment by greg-lunarg** [comment]
@HansKristian-Work  Yes. It seems it was the intent in SPV_EXT_mesh_shader to make this a function terminating instruction ala discard in GLSL or OpKill in SPIR-V, but if so, the specifications and the tool implementations seem to be deficient.

As for GLSL, for some reason I cannot find the spec for GL_EXT_mesh_shader. If I could, I would expect that EmitMeshTasksEXT would be included as a Jump as is `discard`.

The spec you cited for SPV_EXT_mesh_shader also seems deficient. If the intent was to make it a block-ending instruction ala OpKill, it should be included in the set of Function Termination Instructions, but it seems it has not been. Just saying `*OpEmitMeshTasksEXT* must be the last instruction in a block` is vague and in my opinion, insufficient.

And as you noted, if indeed OpEmitMeshTasksEXT is intended as a function terminating instruction, glslang should indeed be generating OpEmitMeshTasksEXT as the last instruction in a block ala OpKill.

@pmistryNV I think you are missing @HansKristian-Work 's point; please see my words above.

I will assign to the author of this PR to glslang.

**Comment by pmistryNV** [comment]
SPV_EXT_mesh_shader clearly states for OpEmitMeshTasksEXT "Ceases all further processing: Only instructions executed before OpEmitMeshTasksEXT have observable side effects."

Hence I think I need to do the following
1. Treat OpEmitMeshTasksEXT as a Function Termination Instruction just like OpTerminateInvocation/OpTerminateRay in glslang
2. Also update the SPV_EXT_mesh_shader spec

**Comment by greg-lunarg** [comment]
In my opinion, the authors of SPV_EXT_mesh_shader @dgkoch and company need to first clarify the intent here: is OpEmitMeshTasksEXT a function termination instruction? If so, GL_EXT_mesh_shader should be updated if needed. Then the tools (glslang, spirv-tools) should be fixed.

**Comment by pmistryNV** [comment]
Yes It is a function termination instruction, I am going to push the spec update

**Comment by pmistryNV** [comment]
@greg-lunarg https://github.com/KhronosGroup/GLSL/pull/190 this is the GLSL spec. 

It states  for EmitMeshTasksEXT
Defines the grid size of subsequent mesh shader workgroups to generate upon completion of the task shader invocation group that called this function and **exits the shader.**

So i think GLSL spec is fine here.

**Comment by sjfricke** [comment]
Note that the `spirv-val` fix is done https://github.com/KhronosGroup/SPIRV-Tools/pull/4923.

@pmistryNV (if you are doing the glslang PR) will need to make sure the `Glsl/CompileToSpirv14Test.FromFile/spv_ext_meshTaskShader_task` and `Glsl/CompileToSpirv14Test.FromFile/spv_460_subgroupEXT_task` test are updated to not include the `OpReturn` after `OpEmitMeshTasksEXT`

**Comment by pmistryNV** [comment]
> Note that the `spirv-val` fix is done [KhronosGroup/SPIRV-Tools#4923](https://github.com/KhronosGroup/SPIRV-Tools/pull/4923).
> 
> @pmistryNV (if you are doing the glslang PR) will need to make sure the `Glsl/CompileToSpirv14Test.FromFile/spv_ext_meshTaskShader_task` and `Glsl/CompileToSpirv14Test.FromFile/spv_460_subgroupEXT_task` test are updated to not include the `OpReturn` after `OpEmitMeshTasksEXT`

Ideally changes should be coordinated and checked in. Since it also needed a spec update and correspondingly needs CTS update there are multiple components that needs to be coordiated. Only check-in spirv-tools caused this breakage
